[Web.JS] call mono_wasm_gc_lock in MonoHeapLock

lambdageek opened the pull request:
Take the runtime's GC lock when creating a heap lock.

**Important Notes**

1. While the runtime GC lock is locked, calling any C# function or doing anything that may trigger managed allocation will block the runtime.
2. In a threaded build, only the main browser thread is allowed to call the `mono_wasm_gc_lock` function
3. The runtime lock is not recursive
4. Background threads initially continue running, so the managed object graph may change even while the GC lock is held (although objects may not move).  Background threads could mutate valuetype fields of reachable objects, even while the GC lock is held. However background threads will block as soon as they try to do any allocation. 
5. The main browser thread may assume that managed heap objects will not move while the lock is held.

Corresponding runtime change dotnet/runtime#82646

lambdageek committed:
call mono_wasm_gc_lock in MonoHeapLock

pavelsavara commented:
Mutating UI components from worker thread is interesting scenario which we should address I think. Otherwise LGTM

lambdageek commented:
Runtime change has been merged.

This PR doesn't depend on the runtime bits (and in fact, won't do anything different until threading is enabled) - if it's paired with a runtime that doesn't have the API, it behaves as before.

This is ready for review now.

SteveSandersonMS commented:
> Background threads initially continue running, so the managed object graph may change even while the GC lock is held (although objects may not move). Background threads could mutate valuetype fields of reachable objects, even while the GC lock is held. However background threads will block as soon as they try to do any allocation.

That's a good point - thanks for raising it. I don't think this can lead to any problems based on how rendering is implemented currently, but we are going to have to make sure we don't later add any logic to `SharedMemoryRenderBatch.ts` that accesses any memory that could mutate due to code isn't on the renderer sync context.

SteveSandersonMS commented:
> Mutating UI components from worker thread is interesting scenario which we should address I think

As long as Blazor WebAssembly implements support for a renderer sync context (equivalent to the one for Blazor Server / WebView), then this should be covered automatically. Components would have to dispatch to the sync context to update their render output, but the sync context would be locked while an existing renderbatch is in process of being applied to the DOM.

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -562,6 +564,10 @@ class MonoHeapLock implements HeapLock {
       throw new Error('Trying to release a lock which isn\'t current');
     }
 
+    if (typeof MONO_INTERNAL !== 'undefined' && typeof MONO_INTERNAL.mono_wasm_gc_unlock === 'function') {
```
I understand this is a point-in-time thing, but maybe we should wait until the runtime is merged before we complete this PR.

It would be ideal *not* to silently skip `mono_wasm_gc_lock`/`unlock` if they can't be found, because then we'll be at risk of accidentally disabling this functionality in the future if the name changes or similar. It would be preferable to fail hard if this function can't be found.

SteveSandersonMS replied:
> if it's paired with a runtime that doesn't have the API, it behaves as before.

I'm not aware that we need to support mismatches of runtime and Blazor versions. Please let me know if I'm missing something!

pavelsavara replied:
Correct, we don't support version mismatch. This is point-in-time comment.

javiercn replied:
If that's the case and we don't want to change it on this PR, can we file an issue to track its removal once the changes are in so that we don't forget?

pavelsavara replied:
I think the whole PR could wait in current state until the runtime version flows.

lambdageek replied:
The runtime changes are in https://github.com/dotnet/aspnetcore/pull/47021

lambdageek replied:
Ok, codeflow from runtime is done.  We're unblocked.  Please take a look again

lambdageek replied:
Or maybe not.  looks like we also need an SDK bump

lambdageek replied:
Or maybe not. looks like we also need an SDK bump

lambdageek replied:
I think we're blocked on https://github.com/dotnet/aspnetcore/pull/47062 which shoudl pull in a runtime that includes at least dotnet/runtime@c7537bdf261 which shoudl have the gc lock API.

I should expect to see failing tests, right?

lambdageek replied:
Looks like we're good now

lambdageek committed:
Depend on the new runtime APIs

blazorwasm and the runtime must be in lockstep. There's no supported
scenario where the API might be missing

lambdageek committed:
don't export MONO_INTERNAL from MonoPlatform.ts

lambdageek commented:
/azp run

lambdageek commented:
/azp run
